### PR TITLE
feat(webapp): add tooltip to main timeline in single view

### PR DESCRIPTION
**Pull request by eh-am**
# Overall
Closes https://github.com/pyroscope-io/pyroscope/issues/1739

Adds a tooltip when hovering over the timeline in SingleView
Screen Shot 2022-11-18 at 15 36 49

# Implementation
Create a `TimelineTooltip` component which just takes the following props:
```ts
export interface TimelineTooltipProps {
  timeLabel: string;
  items: {
    color?: Color;
    value: string;
    label: string;
  }[];
}
```

And renders as is.

Point here is to move the calculations to the caller, since each page may have different values/formats (eg tag explorer).

Also refactors `<ExploreTooltip>` to use this new `TimelineTooltip`.

Ah, fixes https://github.com/pyroscope-io/pyroscope/issues/1740 too

# Some other comments
As it's clear, the Tooltip plugin (from float) has an annoying api, forcing us to access items such as
```
          value: a?.closest?.[1],
```

It's a good idea to refactor that plugin.

### CODE REVIEW

File: webapp/javascript/pages/ContinuousSingleView.tsx [hunk 1]
```diff
@@ -181,6 +188,30 @@ function ContinuousSingleView() {
...
+                );
+
+                if (values.length <= 0) {
+                  return null;
+                }
+
+                return (
+                  <TimelineTooltip timeLabel={data.timeLabel} items={values} />
+                );
+              }}
```
Comment (pavelpashkovsky) [inline webapp/javascript/pages/ContinuousSingleView.tsx:214]: can move this function from JSX to the body of Component?

**Comment by pavelpashkovsky** [comment]
@eh-am can you please test this logic on short timeline period before merging
image

File: webapp/javascript/components/TimelineChart/ExploreTooltip/index.tsx [hunk 2]
```diff
@@ -44,37 +44,36 @@ const ExploreTooltip: FC<ExploreTooltipProps> = ({
...
-        : null}
-    </div>
-  );
+  if (!timeLabel) {
+    return null;
+  }
+
+  if (!values || values.length <= 0) {
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (eh-am) [reply webapp/javascript/components/TimelineChart/ExploreTooltip/index.tsx:66]: Having too many conditions in the same block blurs the boundaries It seems they are connected somehow, which they are not. 

```js
if (!timeLabel || (!values || values.length <= 0))
```

File: webapp/javascript/components/TimelineTooltip.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  return (
+    <div className={styles.valueWrapper}>
+      {ColorDiv}
+      <div>{label}:</div>
+      <div className={styles.closest}>{value}</div>
+    </div>
+  );
+}
+
+export { TimelineTooltip };
```
[the remark this answers is not in the document]
Comment (eh-am) [reply webapp/javascript/components/TimelineTooltip.tsx:54]: That's how the codebase was originally using, but I found that's actually harmful. For more info see https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/

File: webapp/javascript/pages/ContinuousSingleView.tsx [hunk 4]
```diff
@@ -181,6 +188,30 @@ function ContinuousSingleView() {
...
+                  return null;
+                }
+
+                const values = prepareTimelineTooltipContent(
+                  singleView.profile,
+                  query,
+                  data
+                );
+
+                if (values.length <= 0) {
```
[the remark this answers is not in the document]
Comment (eh-am) [reply webapp/javascript/pages/ContinuousSingleView.tsx:207]: Same as above

File: webapp/javascript/pages/ContinuousSingleView.tsx [hunk 5]
```diff
@@ -181,6 +188,30 @@ function ContinuousSingleView() {
               annotations={annotations}
               selectionType="single"
               ContextMenu={contextMenu}
+              onHoverDisplayTooltip={(data) => {
+                if (!singleView || !singleView.profile) {
+                  return null;
+                }
+
+                // TODO: when this could be undefined?
+                if (!data.timeLabel) {
```
[the remark this answers is not in the document]
Comment (eh-am) [reply webapp/javascript/pages/ContinuousSingleView.tsx]: i had a look here and it doesn't seem to happen

File: webapp/javascript/pages/ContinuousSingleView.tsx (hunk 1 above)
Comment (eh-am) [reply webapp/javascript/pages/ContinuousSingleView.tsx]: done

**Pull request merged**